Add support for delegate GDV and method-based vtable GDV

jakobbotsch opened the pull request:
Add support for instrumenting delegate calls and vtable calls into method handle histograms. Use these histograms to do GDV for delegate calls and also support method-based GDV for vtable calls.

For instrumentation we now support class probes at interface call sites, method probes at delegate call sites and both class probes and method probes at vtable call sites. For vtable calls, when turned on, instrumentation produces both histograms as PGO data so that the JIT can later make the choice about what is the best form of guard to use at that site.

For guarding, there are some things to take into account. Delegate calls currently (practically) always point to precode, so this PR is just guarding on `getFunctionFixedEntryPoint` which returns the precode address, and this is generally quite cheap (same cost as class-based GDV). That's the case for delegates pointing to instance methods anyway, this PR does not support static methods yet -- those will be more expensive.

For vtable calls the runtime will backpatch the slots when tiering, so the JIT guards the address retrieved from the vtable against an indirection of the slot, which is slightly more expensive than a class-based guard.

Currently the instrumentation is enabled conditionally with `COMPlus_JitDelegateProfiling=1` (for delegates) and `COMPlus_JitVTableProfiling=1` (for vtable calls). Currently delegate profiling is turned on by default while vtable profiling is off by default.
Simple microbenchmark:
```csharp
public class Benchmark
{
    private readonly long[] _nums;
    public Benchmark()
    {
        _nums = Enumerable.Range(0, 100000).Select(i => (long)i).ToArray();
    }

    [Benchmark]
    public long Sum() => _nums.Sum(l => l * l);
}
```

| Method |        Job |                              Toolchain |      Mean |    Error |   StdDev | Ratio |
|------- |----------- |--------------------------------------- |----------:|---------:|---------:|------:|
|    Sum | Job-QWNDLL |                     \nopgo\corerun.exe | 406.65 us | 0.718 us | 0.560 us |  1.00 |
|    Sum | Job-PNPEDU |                 \tieredpgo_no_delegate_gdv\corerun.exe | 172.77 us | 0.819 us | 0.766 us |  0.42 |
|    Sum | Job-KFFWQK | \tieredpgo_delegate_gdv\corerun.exe |  91.38 us | 0.263 us | 0.219 us |  0.22 |

TODO:
* [X] Work out why chained GDV regresses block layout sometimes (resolved with https://github.com/dotnet/runtime/pull/69878)
* [x] Add to PGO CI pipeline
* [x] Support randomized GDVs for method handle histograms
* [x] Investigate some OSR interactions
* [x] Evaluate proper likelihood heuristics for delegates
* [x] Optimize cold path for delegate GDV -- we should reuse the target
* [x] Devirtualized delegate calls should not null check
* [x] Optimize profiling helpers
* [x] Add signature compatibility check
* [x] Add method address handle comment in disasm (https://github.com/dotnet/runtime/pull/70437)
* [x] Separate ability to do method profiling for delegates and vtable calls
* [ ] ~~Support delegates pointing to value type instance methods~~ (Follow-up)
* [ ] ~~Support delegates pointing to static functions~~ (Follow-up)
* [ ] ~~R2R support~~ (Follow-up)

Contributes to #44610

cc @dotnet/jit-contrib

jakobbotsch committed:
Add support for storing method handle histograms in profiles

Allow method handle histograms in .mibc files and in the PGO text
format.

Contributes to #44610.

jakobbotsch committed:
Fix R2R reader for PGO schema entries without data

jakobbotsch committed:
Support instrumenting method handles for delegates and vtable calls

jakobbotsch committed:
Handle static methods and also move fcall handling

jakobbotsch committed:
Fix writing out methods

jakobbotsch committed:
Remove an unnecessary check

jakobbotsch committed:
Merge remote-tracking branch 'upstream/main' into method-handle-histograms

jakobbotsch committed:
Fix build and some logic inversions

jakobbotsch committed:
Bump R2R minor version and JIT-EE GUID

jakobbotsch committed:
Merge remote-tracking branch 'upstream/main' into method-handle-histograms

jakobbotsch committed:
Merge remote-tracking branch 'upstream/main' into method-handle-instrumentation

jakobbotsch committed:
Merge branch 'method-handle-histograms' into method-handle-instrumentation

jakobbotsch committed:
Merge remote-tracking branch 'origin/method-handle-histograms' into method-handle-instrumentation

jakobbotsch committed:
Fix after merge

jakobbotsch committed:
Change to static_assert and remove a comment

jakobbotsch committed:
Foo

jakobbotsch committed:
Few cleanups

jakobbotsch committed:
Better comment

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
More cleanup

jakobbotsch committed:
Do not null check delegate GDV calls

bencyoung-Fignum commented:
This is amazing. In theory could it hoist the check, clone the loop and then have a loop with a simple add, and a loop with the fallback?

AndyAyersMS commented:
> This is amazing. In theory could it hoist the check, clone the loop and then have a loop with a simple add, and a loop with the fallback?

Yes, we're working on that too. See https://github.com/dotnet/runtime/issues/65206 which includes a link to a prototype that does this (at least for type tests -- should not be hard to make it work for delegate tests as well).

AndyAyersMS commented:
Very cool to see this!  I just skimmed the changes; will take a closer look -- may not get to it until next week.

We should push to get the prereq https://github.com/dotnet/runtime/pull/67919 merged.

> There are some unexpected layout issues due to chained GDV, for example in simple examples `Enumerable.Sum` regresses when method profiling is on because the JIT sees an extra chained GDV opportunity and ends up making a mess of the flow graph.

Not sure what kind of messes you're seeing, but GDV even without chaining can cause problems, eg https://github.com/dotnet/runtime/issues/67318.

> so that the JIT can later make the choice about what is the best form of guard to use at that site.

Would be interesting to see how many sites are type-polymorphic but method-monomorphic.

> both class probes and method probes at vtable call sites.

We should probably work on streamlining the cost of the class/method probes. Simple thing would be to hoist up the check where we decide to update the table; if we're not going to do a table update we don't need to do all the work to figure out what values we'd record. That should help somewhat, especially for the class/method version where those computations are more complex.

jakobbotsch commented:
> We should probably work on streamlining the cost of the class/method probes. Simple thing would be to hoist up the check where we decide to update the table; if we're not going to do a table update we don't need to do all the work to figure out what values we'd record. That should help somewhat, especially for the class/method version where those computations are more complex.

That optimization makes sense. On another note I'm not sure if the combined profile helper is worth much, since from the call site and class we should be able to get the resolved method anyway. So it might just go away in the final version.

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -9536,13 +9536,20 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
             //
             methHnd = callInfo->hMethod;
         }
+        else if (call->AsCall()->IsDelegateInvoke() && opts.OptimizationEnabled() && !opts.IsOSR())
+        {
+            considerGuardedDevirtualization(call->AsCall(), rawILOffset, false, NO_METHOD_HANDLE, NO_CLASS_HANDLE,
+                                            nullptr);
+        }
```
@AndyAyersMS I was seeing issues where we try to both do GDV and add probes in OSR variants. There are comments that say that we currently do not devirtualize in OSR variants here:
https://github.com/dotnet/runtime/blob/66bd8b193bdaa96e26308e6c056f0b27434428cd/src/coreclr/jit/importer.cpp#L20909-L20912

Do you know off the top of your head where we ensure that no GDV should happen for OSR variants?

jakobbotsch replied:
Ah, I think it is the early out of `impDevirtualizeCall` that I have mistakenly removed while extracting `impConsiderCallProbe`.

AndyAyersMS replied:
Yeah, looks like you found it.

jakobbotsch committed:
Avoid devirt for probed calls

jakobbotsch committed:
Cleanup and format

jakobbotsch committed:
Fix 32-bit build

jakobbotsch committed:
Reuse target from temp in cold call

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Fix build

jakobbotsch committed:
Add MethodProfiling PGO scenarios

jakobbotsch committed:
Support randomized GDVs for method handle histograms

jakobbotsch committed:
Run jit-format

jakobbotsch commented:
/azp run runtime-coreclr pgo

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Fix an asertion failure

If we manage to always devirt in the cold path, then we could leave the
control expression causing issues downstream.

jakobbotsch committed:
Make sure we only log in DEBUG

jakobbotsch committed:
Avoid GDV for static calls

jakobbotsch committed:
Fix interaction with chained GDV

For chained GDV candidates, we cannot rely on the 'check' having been
executed, so we cannot reuse the target that was already loaded by the
check.

jakobbotsch committed:
Fix cast helper probes

* Do not insert them when class profiling is turned off
* Use the probe information to determine the probe time. Just
  eligibility is not enough as they are conditionally added.

jakobbotsch committed:
Run jit-format

jakobbotsch commented:
/azp run runtime-coreclr pgo

jakobbotsch committed:
Pick random GDV after logging histogram

jakobbotsch committed:
Fix GDV for vtable calls on System.Delegate

jakobbotsch committed:
Mark early expanded vtable calls as such

On arm32 we may not already have set this flag.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Pull reservoir index calculation up in profile helpers

jakobbotsch committed:
Record only delegate cases that we can actually GDV for

We would record instance methods for open delegate call sites here,
which would result in a mismatch in arg counts during inlining the
devirtualized call. This case is indistinguishable from delegates
pointing at static methods as both require a shuffle thunk that drops
the 'this' arg stored in the delegate, so just record unknown handles
for both these cases. This is not a huge loss as we currently cannot GDV
for delegates pointing to static methods anyway.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Fix pesky x86 GTF_CALL_POP_ARGS bug

We would propagate *all* flags from the control expression into the call
when morphing early-expanded vtable calls. For the cold vtable calls in
the control expression is usually a local variable to reuse the already
loaded call target. We would end up remorphing the call in assertion
prop which happens after liveness that sets GTF_VAR_DEATH on the control
expression. This flag conflicts with GTF_CALL_POP_ARGS that on x86 would
cause us to incorrectly believe the call was caller-cleaned and
unbalance the stack.

jakobbotsch commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

jakobbotsch committed:
Exclude dynamic methods as well as methods in collectible ALCs

jakobbotsch committed:
Remove excessive JIT-EE calls for debug logging in non-verbose

jakobbotsch committed:
Increase some timeouts to get a full test run

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Add missing helper in CorInfoHelpFunc.cs

jakobbotsch committed:
Fix conflict resolution and add profiling helpers on crossgen2 side

jakobbotsch commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

jakobbotsch commented:
> Not sure what kind of messes you're seeing, but GDV even without chaining can cause problems, eg #67318.

Indeed, the case I was seeing look fixed now after #69878. Simple `Enumerable.Sum` benchmark is now significantly faster:
```csharp
public class Benchmark
{
    private readonly long[] _nums;
    public Benchmark()
    {
        _nums = Enumerable.Range(0, 100000).Select(i => (long)i).ToArray();
    }

    [Benchmark]
    public long Sum() => _nums.Sum(l => l * l);
}
```
Results:

| Method |        Job |                              Toolchain |      Mean |    Error |   StdDev | Ratio |
|------- |----------- |--------------------------------------- |----------:|---------:|---------:|------:|
|    Sum | Job-QWNDLL |                     \nopgo\corerun.exe | 406.65 us | 0.718 us | 0.560 us |  1.00 |
|    Sum | Job-PNPEDU |                 \tieredpgo\corerun.exe | 172.77 us | 0.819 us | 0.766 us |  0.42 |
|    Sum | Job-KFFWQK | \tieredpgo_methodprofiling\corerun.exe |  91.38 us | 0.263 us | 0.219 us |  0.22 |

jakobbotsch committed:
Lower likelihood threshold to 30%, like vtable calls

jakobbotsch commented:
I have two test runs where I raised the timeout significantly so that they completed. They are:
runtime-coreclr libraries-pgo: https://dev.azure.com/dnceng/public/_build/results?buildId=1802466&view=results
runtime-coreclr pgo: https://dev.azure.com/dnceng/public/_build/results?buildId=1802465&view=results

There are a few failures but the only failure in the methodprofiling configuration looks to be #70231.

jakobbotsch committed:
Revert "Increase some timeouts to get a full test run"

This reverts commit cbc494b7d53945cd876b20035bb281f81a53109e.

jakobbotsch committed:
Add signature compatibility check

jakobbotsch committed:
Increase some timeouts to get a full test run"

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Fix after merge

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Separate delegate and vtable profiling

jakobbotsch committed:
Set gtTargetHandle for GDV function addrs

jakobbotsch committed:
Fix build

jakobbotsch committed:
Disable method GDV for R2R

jakobbotsch commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

jakobbotsch committed:
Make extra SPMI queries for GDV debugging output purposes

jakobbotsch committed:
Reorder static and signature check

Delegate instances can point to a static method with the first argument
bound; in this case the delegate instance is indistinguishable from one
for an instance method, but the target signature has an extra argument
compared to the call itself.

jakobbotsch committed:
Misc changes

jakobbotsch committed:
Enable delegate profiling by default

jakobbotsch commented:
/azp run runtime-coreclr pgo, runtime-coreclr libraries-pgo

jakobbotsch committed:
Fix newlines when printing test profile

jakobbotsch commented:
I do not see much impact in either direction on TechEmpower:
![image](https://user-images.githubusercontent.com/7887810/173575869-af284281-8f20-4438-ad5c-039afb8b2bb0.png)

That's not that unexpected, we do not use delegates very much in those benchmarks. Here's the data in terms of how many times a probe at a specific IL offset was hit:
https://gist.github.com/jakobbotsch/1ef169922db9be5b3740e05a5ea9544f
For example, the data shows that the function in database-fortunes that has the most collected delegate GDV data was `Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.IOQueue.System.Threading.IThreadPoolWorkItem.Execute`, which had a delegate call site at IL offset 27 that was probed 150243 times (I increased the call counting threshold when gathering the data).

The same data for type GDV is https://gist.github.com/jakobbotsch/0903b426eddb69fd830662b228d5ca32
which has way more data (~12x more lines in that text output) and also the call sites are much hotter.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into method-handle-instrumentation

jakobbotsch committed:
Add GetLikelyMethod and hook it up through R2R

This makes us support generating and consuming the method handle
histograms in the same way as the type handle histograms by compressing
them before they are put in the R2R format.

Also finish some SPMI support.

jakobbotsch commented:
This should be ready. I am enabling delegate GDV by default in this PR, while keeping the vtable profiling disabled. The impact on size of PGO data and SPC.dll with delegate GDV is:

| **File**                      | **Before** | **After** | **PDIFF** |
|-------------------------------|------------|-----------|-----------|
| DotNet_FirstTimeXP.mibc       | 2538 KB    | 2570 KB   | +1.26%    |
| DotNet_HelloWorld.mibc        | 5239 KB    | 5263 KB   | +0.46%    |
| DotNet_OrchardCore.mibc       |  2339 KB   | 2383 KB   | +1.88%    |
| DotNet_TechEmpower.mibc       | 976 KB     | 992 KB    | +1.64%    |
| StandardOptimizationData.mibc | 7377 KB    | 7494 KB   | +1.59%    |
| System.Private.CoreLib.dll        | 11168 KB   | 11168 KB  | +0.00%    |

For TechEmpower impact on PGO, see the comment above.

cc @dotnet/jit-contrib PTAL @AndyAyersMS @EgorBo

jakobbotsch commented on eng/pipelines/common/templates/runtimes/run-test-job.yml:
```diff
@@ -227,13 +227,13 @@ jobs:
     ${{ if in(parameters.testGroup, 'jitstress', 'jitstress-isas-arm', 'jitstressregs-x86', 'jitstressregs', 'jitstress2-jitstressregs', 'gcstress0x3-gcstress0xc', 'ilasm') }}:
       timeoutInMinutes: 390
-    ${{ if in(parameters.testGroup, 'gcstress-extra', 'r2r-extra', 'clrinterpreter') }}:
+    ${{ if in(parameters.testGroup, 'gcstress-extra', 'r2r-extra', 'clrinterpreter', 'pgo') }}:
       timeoutInMinutes: 510
```
I am thinking to keep these timeout changes until we get around to splitting up the jobs, it fixes the problem we are seeing with timeouts and I have not seen any issues with making the timeout this high.

jakobbotsch committed:
Update JIT-EE GUID

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -21389,51 +21405,7 @@ void Compiler::impDevirtualizeCall(GenTreeCall*            call,
-    //
-    // We strip BBINSTR from inlinees currently, so we'll only
-    // do this for the root method calls.
-    //
-    if (opts.jitFlags->IsSet(JitFlags::JIT_FLAG_BBINSTR))
-    {
-        assert(opts.OptimizationDisabled() || opts.IsOSR());
-        assert(!compIsForInlining());
-
-        // During importation, optionally flag this block as one that
-        // contains calls requiring class profiling. Ideally perhaps
-        // we'd just keep track of the calls themselves, so we don't
-        // have to search for them later.
-        //
-        if ((call->gtCallType != CT_INDIRECT) && opts.jitFlags->IsSet(JitFlags::JIT_FLAG_BBINSTR) &&
-            !opts.jitFlags->IsSet(JitFlags::JIT_FLAG_PREJIT) && (JitConfig.JitClassProfiling() > 0) &&
-            !isLateDevirtualization)
-        {
-            JITDUMP("\n ... marking [%06u] in " FMT_BB " for class profile instrumentation\n", dspTreeID(call),
-                    compCurBB->bbNum);
-            ClassProfileCandidateInfo* pInfo = new (this, CMK_Inlining) ClassProfileCandidateInfo;
-
-            // Record some info needed for the class profiling probe.
-            //
-            pInfo->ilOffset                   = ilOffset;
-            pInfo->probeIndex                 = info.compClassProbeCount++;
-            call->gtClassProfileCandidateInfo = pInfo;
-
-            // Flag block as needing scrutiny
-            //
-            compCurBB->bbFlags |= BBF_HAS_CLASS_PROFILE;
-        }
-        return;
-    }
-
-    // Bail if optimizations are disabled.
-    if (opts.OptimizationDisabled())
-    {
-        return;
-    }
```
This code was extracted into impConsiderCallProbe which is called in impImportCall instead. Unlike impDevirtualizeCall, impConsiderCallProbe is called for delegates too.

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -22225,153 +22305,378 @@ void Compiler::addFatPointerCandidate(GenTreeCall* call)
-    dvInfo.context                     = *pContextHandle;
-    dvInfo.exactContext                = *pContextHandle;
-    dvInfo.pResolvedTokenVirtualMethod = nullptr;
+    uint32_t likelyClassAttribs = 0;
+    if (likelyClass != NO_CLASS_HANDLE)
+    {
+        likelyClassAttribs = info.compCompHnd->getClassAttribs(likelyClass);
```
This stuff is easier to review with whitespace ignored.

jakobbotsch commented:
Test failure is #68376. superpmi failures are expected due to JIT-EE GUID change.

AndyAyersMS reviewed (commented):
Overall this looks great.

jakobbotsch committed:
Address some feedback, remove some unneeded stuff

jakobbotsch commented on src/coreclr/jit/compiler.h:
```diff
@@ -6795,13 +6811,26 @@ class Compiler
         optMethodFlags |= OMF_HAS_GUARDEDDEVIRT;
     }
 
+    void clearMethodHasGuardedDevirtualization()
+    {
+        optMethodFlags &= ~OMF_HAS_GUARDEDDEVIRT;
+    }
+
```
Ah, I was wondering why I added this. Looks like a bad merge. Will fix.

jakobbotsch commented on src/coreclr/jit/fgprofile.cpp:
```diff
@@ -1440,89 +1440,101 @@ class ClassProbeVisitor final : public GenTreeVisitor<ClassProbeVisitor<TFunctor
+        }
+    }
+
+    void CreateHistogramEntriesIfNecessary(Compiler* compiler, GenTreeCall* call, bool isTypeHistogram)
```
Hmm, I think at one point I was checking the probe type in that function but decided to do the current thing. Will rename it.

jakobbotsch committed:
Address some more feedback

jakobbotsch commented:
Also PTAL @davidwrighton for the non-JIT/SPMI changes. I have bumped the R2R minor version again since we now may see `GetLikelyMethod` entries in the embedded PGO data.

One thing I would like to point attention to is that old versions of dotnet-pgo (from before #67919) fail with an "Unknown PGO type" error if they see this new data. So one needs to use a version of dotnet-pgo that matches the version of the runtime used to produce the data.

davidwrighton reviewed (approved):
:shipit:

davidfowl commented:
I missed this change. This is a big deal 😀 !
